### PR TITLE
[opt](file cache) make the cache queue percent configurable

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/common/config.cpp [hunk 1]
```diff
@@ -999,6 +999,7 @@ DEFINE_Validator(file_cache_min_file_segment_size, [](const int64_t config) -> b
 DEFINE_Bool(clear_file_cache, "false");
 DEFINE_Bool(enable_file_cache_query_limit, "false");
 DEFINE_mInt32(file_cache_wait_sec_after_fail, "0"); // // zero for no waiting and retrying
+DEFINE_String(file_cache_queue_percentage, "[17,2,1,20]");
```
Comment (xiaokang) [inline be/src/common/config.cpp:1002]: suggest name file_cache_queue_percent

Comment (xiaokang) [reply be/src/common/config.cpp:1002]: 17 + 2 + 1 + 20 < 100

Comment (xiaokang) [reply be/src/common/config.cpp:1002]: add doc for be config

File: be/src/olap/options.cpp [hunk 2]
```diff
@@ -244,18 +244,44 @@ Status parse_conf_cache_paths(const std::string& config_path, std::vector<CacheP
     return Status::OK();
 }
 
-io::FileCacheSettings CachePath::init_settings() const {
+Status parse_conf_cache_queue_percentage(std::string percentage_config,
+                                         std::vector<size_t>& percentages) {
+    if (percentage_config.front() != '[' || percentage_config.back() != ']') {
+        return Status::InvalidArgument("wrong percentage format");
```
Comment (xiaokang) [inline be/src/olap/options.cpp:250]: add more info for valid format

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/options.cpp [hunk 3]
```diff
@@ -235,7 +236,30 @@ Status parse_conf_cache_paths(const std::string& config_path, std::vector<CacheP
...
+                    }
+                }
+            } else {
+                return Status::InvalidArgument("percents should be array");
+            }
+        } else {
+            percents.assign({85, 10, 5});
+        }
+        if (percents.size() != 3 || percents[0] + percents[1] + percents[2] != 100) {
+            return Status::InvalidArgument("wrong percents value");
```
Comment (Lchangliang) [inline be/src/olap/options.cpp:259]: More detail. The array size is not equal to 3 or the sum is not equal to 100.

File: be/src/olap/options.cpp [hunk 4]
```diff
@@ -244,18 +268,18 @@ Status parse_conf_cache_paths(const std::string& config_path, std::vector<CacheP
     return Status::OK();
 }
 
-io::FileCacheSettings CachePath::init_settings() const {
+io::FileCacheSettings CachePath::init_settings(const std::vector<size_t>& percents) const {
```
Comment (Lchangliang) [inline be/src/olap/options.cpp:271]: The percents is member variable. Not need the input param.

File: be/src/olap/options.cpp [hunk 5]
```diff
@@ -200,7 +201,7 @@ void parse_conf_broken_store_paths(const string& config_path, std::set<std::stri
  *  [
  *    {"path": "storage1", "total_size":53687091200,"query_limit": "10737418240"},
  *    {"path": "storage2", "total_size":53687091200},
- *    {"path": "storage3", "total_size":53687091200},
+ *    {"path": "storage3", "total_size":53687091200, percents: [85,10,5]}
```
Comment (Lchangliang) [inline be/src/olap/options.cpp:204]: Modify the annotation in "config.h" too

File: be/test/io/cache/file_block_cache_test.cpp [hunk 6]
```diff
@@ -118,7 +118,7 @@ TEST(LRUFileCache, init) {
     EXPECT_TRUE(parse_conf_cache_paths(string, cache_paths));
     EXPECT_EQ(cache_paths.size(), 2);
     for (const auto& cache_path : cache_paths) {
-        io::FileCacheSettings settings = cache_path.init_settings();
+        io::FileCacheSettings settings = cache_path.init_settings(cache_path.percents);
```
Comment (Lchangliang) [inline be/test/io/cache/file_block_cache_test.cpp:121]: Add some ut to cover it.

**Comment by zzzxl1993** [comment]
run buildall

**Review by Lchangliang — APPROVED** [review]

**Comment by Lchangliang** [comment]
LGTM

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/options.cpp [hunk 7]
```diff
@@ -235,7 +238,38 @@ Status parse_conf_cache_paths(const std::string& config_path, std::vector<CacheP
...
+        bool has_normal_percent = map.HasMember(CACHE_NORMAL_PERCENT.c_str());
+        bool has_disposable_percent = map.HasMember(CACHE_DISPOSABLE_PERCENT.c_str());
+        bool has_index_percent = map.HasMember(CACHE_INDEX_PERCENT.c_str());
+        if (has_normal_percent && has_disposable_percent && has_index_percent) {
+            RETURN_IF_ERROR(get_percent_value(CACHE_NORMAL_PERCENT, normal_percent));
+            RETURN_IF_ERROR(get_percent_value(CACHE_DISPOSABLE_PERCENT, disposable_percent));
+            RETURN_IF_ERROR(get_percent_value(CACHE_INDEX_PERCENT, index_percent));
+        } else if (has_normal_percent || has_disposable_percent || has_index_percent) {
+            return Status::InvalidArgument(
+                    "percent configurations must either be all set or all unset.");
```
Comment (xiaokang) [inline be/src/olap/options.cpp:265]: cache percent config

File: be/src/olap/options.cpp [hunk 8]
```diff
@@ -235,7 +238,38 @@ Status parse_conf_cache_paths(const std::string& config_path, std::vector<CacheP
...
+        }
+        if ((normal_percent + disposable_percent + index_percent) != 100) {
+            return Status::InvalidArgument("the sum is not equal to 100.");
```
Comment (xiaokang) [inline be/src/olap/options.cpp:268]: sum of cache percent config

**Comment by zzzxl1993** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Comment by zzzxl1993** [comment]
run pipelinex_p0

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**